### ISSUE TITLE
Need help to understand sequence level unlikelihood loss function.

### ISSUE THREAD

**Issue by Sleepyhead01** [facebookresearch/unlikelihood_training#10]
https://github.com/facebookresearch/unlikelihood_training/blob/944465589c0fab534fe6d14a5db2850ddeee43ce/custom/gpt2/run_gpt2.py#L131

```
    input_sequence = batch[0].cuda()
    batch = batch_input_sequence_by_prefix_length(input_sequence, args.prefix_length)
    completions, continuation_logits = sample_sequence(model, batch, args.prefix_length, args.continuation_length, args.top_k, args.top_p)
    pred_toks = completions[:, args.prefix_length:].contiguous()

    mask = ngram_repeat_mask(pred_toks, args.sequence_ngram_n).type_as(continuation_logits)

    lprobs = F.log_softmax(continuation_logits, dim=-1)
    pred_lprobs = lprobs.view(-1, lprobs.size(2)).gather(1, pred_toks.view(-1, 1))
    one_minus_probs = torch.clamp((1.0 - pred_lprobs.exp()), min=1e-20).view(pred_toks.size(0), pred_toks.size(1))
    loss = -torch.log(one_minus_probs) * mask
    loss = loss.sum()
    ntokens = pred_toks.numel()  # number of output tokens (tokens in completions)
```

What exactly do the variables completions, continuation_logits, and mask represent according to the paper? Would appreciate it a lot if someone can explain how the custom loss is calculated here.

**Comment by uralik** [comment]
Hi @Sleepyhead01 ! 

* completions - these are sequences **including the context** generated by the model at the given training step. In the paper terms, this is a 'completion', see sequence completion paragraph in section 3. In short, completion = context + continuation, where + is a concat operation.

* continuation_logits - these are logits of the continuation only (so not including the context).

* mask - this is a binary mask where 1 means that the given token is in the repeated ngram and 0 otherwise. This mask is used later to only keep the repeated tokens in the objective.

I hope it is clear now, please ask more questions if anything else is not clear!
